File: Week_04-BFS-DFS-二分-贪心/200.岛屿数量.cpp
/*
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * [200] 岛屿数量
 */

// @lc code=start

class Solution {
#if 1
/*
方法一，深度优先搜索
扫描整个而为网格，如果某个位置为1，把它作为起点开始进行深度优先搜索。
搜索过程中，每找到一个1，在原网格中，把它标记为0.

最终岛屿数量，就是进行深度优先搜索的次数
*/
private:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        int nr = grid.size();
        int nc = grid[0].size();
        grid[r][c] = '0';
        if (r - 1 >= 0 && grid[r-1][c] == '1') dfs(grid, r - 1, c);
        if (r + 1 < nr && grid[r+1][c] == '1') dfs(grid, r + 1, c);
        if (c - 1 >= 0 && grid[r][c-1] == '1') dfs(grid, r, c - 1);
        if (c + 1 < nc && grid[r][c+1] == '1') dfs(grid, r, c + 1);
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int nr = grid.size();
        if (!nr) return 0;
        int nc = grid[0].size();
        int num_islands = 0;
        for (int r = 0; r < nr; r++) {
            for (int c = 0; c < nc; c++) {
                if(grid[r][c] == '1') {
                    num_islands++;
                    dfs(grid, r, c);
                }
            }
        }
        return num_islands;
    }
#endif
};
// @lc code=end

File: Week_04-BFS-DFS-二分-贪心/200.岛屿数量.py
#
# @lc app=leetcode.cn id=200 lang=python
#
# [200] 岛屿数量
#

# @lc code=start
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        def dfs(grid, r, c):
            nr = len(grid)
            nc = len(grid[0])
            grid[r][c] = '0'
            if r-1>=0 and grid[r-1][c] == '1': dfs(grid, r-1, c)
            if r+1<nr and grid[r+1][c] == '1': dfs(grid, r+1, c)
            if c-1>=0 and grid[r][c-1] == '1': dfs(grid, r, c-1)
            if c+1<nc and grid[r][c+1] == '1': dfs(grid, r, c+1)

        nr = len(grid)
        if nr == 0: return 0
        nc = len(grid[0])
        num_island = 0
        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == '1':
                    num_island += 1
                    dfs(grid, r, c)
        return num_island
# @lc code=end

